### PR TITLE
Adds the cargo transporter

### PR DESCRIPTION
**Pull request by j-mie**
:cl: JamieH
add: Added the cargo transporter, inspired/ported from Goon.
add: You can use it to teleport various different items around the station, cargo and mining start with the cargo transported and they have some spares lying around too.
/:cl:

Adds in a cargo transporter, inspiration/some code from Goon.

Cargo Telepads are located at:
- Cargo Bay
- Engineering
- Medbay
- Research
- Mining Output

Look here: https://share.jamiehankins.co.uk/ShareX/2017/02/2017-02-17_09-05-38.mp4
(I fixed the bug at the end)

### CODE REVIEW

**Comment by ghost** [comment]
Dirty branch. Otherwise the RCS was a good feature back from pre-pre-rebase. +1

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 1]
```diff
@@ -240,6 +240,9 @@
 							"<span class='notice'>You [welded ? "weld" : "unwelded"] \the [src] with \the [WT].</span>",
 							"<span class='italics'>You hear welding.</span>")
 			update_icon()
+	if (istype(W, /obj/item/weapon/rcs))
+		W:cargoteleport(src, user)
```
Comment (ghost) [inline code/game/objects/structures/crates_lockers/closets.dm]: **NO**

**_DO NOT USE THE COLON CHARACTER_**

File: code/modules/atmospherics/machinery/portable/canister.dm [hunk 2]
```diff
@@ -198,7 +198,10 @@
 	qdel(src)
 
 /obj/machinery/portable_atmospherics/canister/attackby(obj/item/weapon/W, mob/user, params)
-	if(user.a_intent != INTENT_HARM && istype(W, /obj/item/weapon/weldingtool))
+	if(user.a_intent != INTENT_HARM && istype(W, /obj/item/weapon/rcs))
+		W:cargoteleport(src, user)
```
Comment (ghost) [inline code/modules/atmospherics/machinery/portable/canister.dm]: **NO**

**_DO NOT USE THE COLON CHARACTER_**

File: code/modules/mining/satchel_ore_boxdm.dm [hunk 3]
```diff
@@ -25,6 +25,9 @@
 		if(do_after(user, 50*C.toolspeed, target = src))
 			user.visible_message("[user] pries \the [src] apart.", "<span class='notice'>You pry apart \the [src].</span>", "<span class='italics'>You hear splitting wood.</span>")
 			deconstruct(TRUE, user)
+	else if(istype(W, /obj/item/weapon/rcs))
+		W:cargoteleport(src, user)
```
Comment (ghost) [inline code/modules/mining/satchel_ore_boxdm.dm]: **NO**

**_DO NOT USE THE COLON CHARACTER_**

File: code/modules/telesci/telepad.dm [hunk 4]
```diff
@@ -51,18 +51,33 @@
 
 	return ..()
 
+var/global/list/cargopads = list() // Global List of Cargo Pads
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: Put this in `code/_globalvars`

File: code/modules/telesci/telepad.dm [hunk 5]
```diff
@@ -51,18 +51,33 @@
 
 	return ..()
 
+var/global/list/cargopads = list() // Global List of Cargo Pads
 
 //CARGO TELEPAD//
 /obj/machinery/telepad_cargo
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: Create a new telepad object and put it in the hippiestation module, please leave the current code unchanged for the sake of easier upstream merges

File: code/modules/telesci/telepad.dm [hunk 6]
```diff
@@ -51,18 +51,33 @@
...
+	desc = "A telepad used by the cargo transporter."
 	icon = 'icons/obj/telescience.dmi'
-	icon_state = "pad-idle"
+	icon_state = "pad-idle-o"
 	anchored = 1
 	use_power = 1
 	idle_power_usage = 20
 	active_power_usage = 500
 	var/stage = 0
+	var/active = 0
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: `FALSE`

File: code/modules/telesci/telepad.dm [hunk 7]
```diff
@@ -51,18 +51,33 @@
...
 
 //CARGO TELEPAD//
 /obj/machinery/telepad_cargo
 	name = "cargo telepad"
-	desc = "A telepad used by the Rapid Crate Sender."
+	desc = "A telepad used by the cargo transporter."
 	icon = 'icons/obj/telescience.dmi'
-	icon_state = "pad-idle"
+	icon_state = "pad-idle-o"
 	anchored = 1
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: `TRUE`

File: code/modules/telesci/telepad.dm [hunk 8]
```diff
@@ -51,18 +51,33 @@
...
 	use_power = 1
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: `TRUE`

File: code/modules/telesci/telepad.dm [hunk 9]
```diff
@@ -51,18 +51,33 @@
...
 	idle_power_usage = 20
 	active_power_usage = 500
 	var/stage = 0
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: No magic numbers please

File: code/modules/telesci/telepad.dm [hunk 10]
```diff
@@ -51,18 +51,33 @@
...
+	var/active = 0
+
+/obj/machinery/telepad_cargo/New()
+	..()
+	if (name == "cargo telepad")
+		name += " ([rand(100,999)])"
+	if (active && !cargopads.Find(src))
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: Please use the list helpers when accessing lists, `Find` is rather slow and there are better solutions as seen here: https://github.com/tgstation/tgstation/blob/c30d07f55a1078b6463b56046884fbce9bac6959/code/__HELPERS/lists.dm

File: code/modules/telesci/telepad.dm [hunk 11]
```diff
@@ -51,18 +51,33 @@
...
+		cargopads.Add(src)
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: Please use the list helpers when accessing lists, `Add` is rather slow and there are better solutions as seen here: https://github.com/tgstation/tgstation/blob/c30d07f55a1078b6463b56046884fbce9bac6959/code/__HELPERS/lists.dm

File: code/modules/telesci/telepad.dm [hunk 12]
```diff
@@ -51,18 +51,33 @@
...
+
+/obj/machinery/telepad_cargo/Destroy()
+	if (cargopads.Find(src))
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: Please use the list helpers when accessing lists, `Find` is rather slow and there are better solutions as seen here: https://github.com/tgstation/tgstation/blob/c30d07f55a1078b6463b56046884fbce9bac6959/code/__HELPERS/lists.dm

File: code/modules/telesci/telepad.dm [hunk 13]
```diff
@@ -51,18 +51,33 @@
...
+		cargopads.Remove(src)
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: Please use the list helpers when accessing lists, `Remove` is rather slow and there are better solutions as seen here: https://github.com/tgstation/tgstation/blob/c30d07f55a1078b6463b56046884fbce9bac6959/code/__HELPERS/lists.dm

File: code/modules/telesci/telepad.dm [hunk 14]
```diff
@@ -97,6 +112,20 @@
...
+		icon_state = "pad-idle-o"
+		active = 0
+		if (cargopads.Find(src))
+			cargopads.Remove(src)
+	else
+		user << "You switch the receiver on."
+		icon_state = "pad-idle"
+		active = 1
+		if (!cargopads.Find(src))
+			cargopads.Add(src)
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: Please use the list helpers when accessing lists, `Add` is rather slow and there are better solutions as seen here: https://github.com/tgstation/tgstation/blob/c30d07f55a1078b6463b56046884fbce9bac6959/code/__HELPERS/lists.dm

File: code/modules/telesci/telepad.dm [hunk 15]
```diff
@@ -97,6 +112,20 @@
...
+		user << "You switch the receiver off."
+		icon_state = "pad-idle-o"
+		active = 0
+		if (cargopads.Find(src))
+			cargopads.Remove(src)
+	else
+		user << "You switch the receiver on."
+		icon_state = "pad-idle"
+		active = 1
+		if (!cargopads.Find(src))
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: Please use the list helpers when accessing lists, `Find` is rather slow and there are better solutions as seen here: https://github.com/tgstation/tgstation/blob/c30d07f55a1078b6463b56046884fbce9bac6959/code/__HELPERS/lists.dm

File: code/modules/telesci/telepad.dm [hunk 16]
```diff
@@ -97,6 +112,20 @@
 	else
 		return ..()
 
+/obj/machinery/telepad_cargo/attack_hand(mob/living/user)
+	if (active == 1)
+		user << "You switch the receiver off."
+		icon_state = "pad-idle-o"
+		active = 0
+		if (cargopads.Find(src))
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: Please use the list helpers when accessing lists, `Find` is rather slow and there are better solutions as seen here: https://github.com/tgstation/tgstation/blob/c30d07f55a1078b6463b56046884fbce9bac6959/code/__HELPERS/lists.dm

File: code/modules/telesci/telepad.dm [hunk 17]
```diff
@@ -97,6 +112,20 @@
 	else
 		return ..()
 
+/obj/machinery/telepad_cargo/attack_hand(mob/living/user)
+	if (active == 1)
+		user << "You switch the receiver off."
+		icon_state = "pad-idle-o"
+		active = 0
+		if (cargopads.Find(src))
+			cargopads.Remove(src)
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: Please use the list helpers when accessing lists, `Remove` is rather slow and there are better solutions as seen here: https://github.com/tgstation/tgstation/blob/c30d07f55a1078b6463b56046884fbce9bac6959/code/__HELPERS/lists.dm

File: code/modules/telesci/telepad.dm [hunk 18]
```diff
@@ -97,6 +112,20 @@
 	else
 		return ..()
 
+/obj/machinery/telepad_cargo/attack_hand(mob/living/user)
+	if (active == 1)
+		user << "You switch the receiver off."
+		icon_state = "pad-idle-o"
+		active = 0
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: `FALSE`

File: code/modules/telesci/telepad.dm [hunk 19]
```diff
@@ -97,6 +112,20 @@
...
+	if (active == 1)
+		user << "You switch the receiver off."
+		icon_state = "pad-idle-o"
+		active = 0
+		if (cargopads.Find(src))
+			cargopads.Remove(src)
+	else
+		user << "You switch the receiver on."
+		icon_state = "pad-idle"
+		active = 1
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: `TRUE`

**Review by ghost — CHANGES_REQUESTED** [review]

File: code/modules/telesci/telepad.dm [hunk 20]
```diff
@@ -116,59 +145,85 @@
...
+
+	var/charges = 10
+	var/maximum_charges = 10.0
+	var/obj/machinery/telepad_cargo/target_pad = null
+	var/target_turf = null
 
 /obj/item/weapon/rcs/examine(mob/user)
 	..()
-	user << "There are [rcharges] charge\s left."
+	user << "There are [charges]/[maximum_charges] charge\s left."
```
Comment (ghost) [inline code/modules/telesci/telepad.dm:170]: This doesn't make sense? This is gonna be 0 to 1 charges left depending on the value.

Comment (ghost) [reply code/modules/telesci/telepad.dm:170]: Unless you want this to be a percentage in which you should do the *100 and capping the amount of decimal places and putting a percent sign after the calculation.

File: code/modules/telesci/telepad.dm [hunk 21]
```diff
@@ -116,59 +145,85 @@
...
-		if(mode == 0)
-			mode = 1
-			playsound(src.loc, 'sound/effects/pop.ogg', 50, 0)
-			user << "<span class='caution'>The telepad locator has become uncalibrated.</span>"
+	user << "<span style=\"color:blue\">Teleporting [T]...</span>"
+	playsound(user.loc, "sound/machines/click.ogg", 50, 1)
+
+	target_pad.icon_state = "pad-beam"
+
+	if(do_after(user, 50, target=T, progress = 1))
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: `do_after` can be stacked and you only remove charges after `do_after` has finished, allowing the user to use the item more often

File: code/modules/telesci/telepad.dm [hunk 22]
```diff
@@ -116,59 +145,85 @@
...
 ///HANDHELD TELEPAD USER///
 /obj/item/weapon/rcs
-	name = "rapid-crate-sender (RCS)"
-	desc = "Use this to send crates and closets to cargo telepads."
+	name = "cargo transporter"
+	desc = "A device for teleporting crated goods."
 	icon = 'icons/obj/telescience.dmi'
 	icon_state = "rcs"
 	flags = CONDUCT
 	force = 10
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: why does this have a force of 10?

File: code/modules/telesci/telepad.dm [hunk 23]
```diff
@@ -116,59 +145,85 @@
...
 	throwforce = 10
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: why does this have a throwforce of 10?

File: code/modules/telesci/telepad.dm [hunk 24]
```diff
@@ -97,6 +112,20 @@
 	else
 		return ..()
 
+/obj/machinery/telepad_cargo/attack_hand(mob/living/user)
+	if (active == 1)
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: `if(active)`

File: code/modules/telesci/telepad.dm [hunk 25]
```diff
@@ -51,18 +51,33 @@
...
+		cargopads.Add(src)
+
+/obj/machinery/telepad_cargo/Destroy()
+	if (cargopads.Find(src))
+		cargopads.Remove(src)
+	..()
+
 /obj/machinery/telepad_cargo/attackby(obj/item/weapon/W, mob/user, params)
 	if(istype(W, /obj/item/weapon/wrench))
 		anchored = 0
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: `= FALSE`

File: code/game/machinery/recharger.dm [hunk 26]
```diff
@@ -142,6 +142,13 @@
 						use_power(200 * recharge_coeff)
 						using_power = 1
 
+		if(istype(charging, /obj/item/weapon/rcs))
+			var/obj/item/weapon/rcs/C = charging
+			if (C.charges < C.maximum_charges)
+				C.charges++
+				use_power(250 * recharge_coeff)
+				using_power = 1
```
Comment (ghost) [inline code/game/machinery/recharger.dm:150]: `TRUE`

File: code/modules/telesci/telepad.dm (hunk 20 above)
Comment (j-mie) [reply code/modules/telesci/telepad.dm:170]: It's text, not math, so it says "1/10 charges left" or "4/10 charges left"

File: code/modules/telesci/telepad.dm (hunk 22 above)
Comment (j-mie) [reply code/modules/telesci/telepad.dm]: fixed

File: code/game/machinery/recharger.dm (hunk 26 above)
Comment (j-mie) [reply code/game/machinery/recharger.dm:150]: The whole proc uses 1s and 0s, it'll just look weird.

File: code/modules/telesci/telepad.dm [hunk 27]
```diff
@@ -51,27 +51,41 @@
...
+	icon_state = "pad-idle-o"
+	anchored = TRUE
+	use_power = TRUE
 	idle_power_usage = 20
 	active_power_usage = 500
+
 	var/stage = 0
+	var/active = FALSE
+
+/obj/machinery/telepad_cargo/New()
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: /tg/ added initialize for station objects, mind taking a look if that might apply here?

File: code/modules/telesci/telepad.dm [hunk 28]
```diff
@@ -51,27 +51,41 @@
...
+	..()
+	if (name == "cargo telepad")
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: `if == initial(src.name)`

File: code/modules/telesci/telepad.dm [hunk 29]
```diff
@@ -97,6 +111,20 @@
 	else
 		return ..()
 
+/obj/machinery/telepad_cargo/attack_hand(mob/living/user)
```
Comment (Chnkr) [inline code/modules/telesci/telepad.dm]: add a do_after in this proc to prevent players from crashing the server

Comment (Chnkr) [inline code/modules/telesci/telepad.dm]: or just a `CHECK_TICK` really, list operations in BYOND are terrible

File: code/modules/telesci/telepad.dm [hunk 30]
```diff
@@ -112,18 +112,20 @@
...
-		active = FALSE
-		if (cargopads[src] != null)
-			LAZYREMOVE(cargopads, src)
-	else
-		user << "You switch the receiver on."
-		icon_state = "pad-idle"
-		active = TRUE
-		if (cargopads[src] != null)
-			LAZYADD(cargopads, src)
+	if(do_after(user, 20*W.toolspeed, target = src, unique=TRUE))
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: attack_hand has no toolspeed

Comment (ghost) [reply code/modules/telesci/telepad.dm]: also the do_after is not needed since we have logging, and it's shit to do a do_after for something so simple

**Comment by KayeArray** [comment]
Hell yes

File: code/modules/telesci/telepad.dm [hunk 31]
```diff
@@ -112,7 +113,7 @@
 		return ..()
 
 /obj/machinery/telepad_cargo/attack_hand(mob/living/user)
-	if(do_after(user, 20*W.toolspeed, target = src, unique=TRUE))
+	if(do_after(user, 5, target = src))
```
Comment (ghost) [inline code/modules/telesci/telepad.dm]: needs a visible message

**Comment by tortellinitony** [comment]
We have MULEbots for a reason, come on

**Comment by ghost** [comment]
@tortellinitony mulebots are slow, maybe we should allow RD to upgrade mulebots?

File: code/modules/telesci/telepad.dm (hunk 20 above)
Comment (ghost) [reply code/modules/telesci/telepad.dm:170]: OH, sorry, I thought the whole expression was wrapped in square brackets, my bad.

**Comment by tortellinitony** [comment]
there's literally a turbo speed wire @Chnkr

**Comment by cacogen** [comment]
so they can be recharged? does cargo get a recharger though, or will they just be breaking into the cargo checkpoint?

also this is really good, moving crates is such a pain

**Review by tortellinitony — APPROVED** [review]

**Pull request merged**